src/SkyHope.LibraryManager.WebApi/Controllers/BookController.cs

```diff
@@ -1,12 +1,258 @@
+            _logger = logger;
+            _logger.LogDebug(1, "NLog injected into BookController");
+        }
 
+        [HttpGet]
+        public async Task<ActionResult<IEnumerable<HttpBook>>> GetAllAsync()
+        {
+            var results = new List<HttpBook>();
+            var allBooks = await _bookRepository.GetAllAsync();
+            var availableBooks = allBooks.Where(b => b.IsAvailable && !b.IsDeleted).ToList();
+
+            if (!availableBooks.Any())
+            {
+                return NotFound();
+            }
+
+            foreach (var book in availableBooks)
+            {
+                results.Add(new HttpBook
+                {
+                    Title = book.Title,
+                    BookId = book.BookId,
+                    AuthorId = book.AuthorId,
+                    DewyClass = book.DewyClass,
+                    Isbn = book.Isbn,
+                    Year = book.Year
+                });
+            }
+            _logger.LogInformation($"GetAllAsync in BooksController returned {availableBooks.Count} available books.");
+            return Ok(results);
+        }
+
+        [HttpGet]
+        public async Task<ActionResult<HttpBook>> GetByIdAsync(int id)
+        {
+            var book = await _bookRepository.GetEntityAsync(id);
+            if (book == null)
+            {
+                return NotFound();
+            }
+
+            var result = new HttpBook
+            {
+                Title = book.Title,
+                BookId = book.BookId,
+                AuthorId = book.AuthorId,
+                DewyClass = book.DewyClass,
+                Isbn = book.Isbn,
+                Year = book.Year
+            };
+
+            return Ok(result);
+        }
+
+        [HttpPatch]
+        public async Task<ActionResult> DeleteAsync(HttpBook book)
+        {
+            if (!await _bookRepository.TryDeleteAsync(book.BookId))
+            {
+                return StatusCode(StatusCodes.Status500InternalServerError, "An unexpected error occurred.");
+            }
+            return Ok();
+        }
+
+        [HttpPut]
+        public async Task<ActionResult> CheckOutAsync(HttpBook book, HttpModels.User user)
+        {
+            var bookToUpdate = await _bookRepository.GetEntityAsync(book.BookId);
+            if (bookToUpdate is null)
+            {
+                return NotFound();
+            }
+
+            if (!bookToUpdate.IsAvailable)
+            {
+                return BadRequest("This book is not available for checkout.");
+            }
+
+            var userToAssign = await _userRepository.GetEntityAsync(user.UserId);
+            if (userToAssign is null)
+            {
+                return NotFound("Could not find user to assign to book.");
+            }
+
+            if (userToAssign.CheckedOutBooks.Count >= maxBookCount)
+            {
+                return BadRequest($"This user has {maxBookCount} books checked out, and may not check out more.");
```

Comment:
I wouldn't make this a bad request. There's nothing wrong with the actual request here. The problem isn't a technical one - it's a business one, therefore I'd return OK but but make a return type (a class) that can encapsulate the results that you might expect (i.e. "Success", "Already Checked Out", "Late Fees Due")